Feature/power&bms

for-a-greener-dawn opened the pull request:
This pull request of the Feature/power&bms contains the Power&BMS.tsx file belonging in the components folder of RED. The purpose of the Power and BMS feature is to receive data collected by the rover about various components' power draw and battery statuses, to send commands to enable/disable certain power busses, and to send a command to either power cycle the rover or turn it off. No known file conflicts exist at the writing of this request.

This branch was authored by @Machtri with significant assistance from @thomas-gleiforst and @amwgvf 

This project, sans non-essential color converting for the data containers, does not have any file dependencies apart from the files required by most of the components files.

This PR contains:

- addition of Power&BMS.tsx to Electron/RED/components/
- alter to ControlCenter.tsx in Electron/RED/ to import Power&BMS and insert <Power> in between <GPS> and <Log>

for-a-greener-dawn committed:
Create Power&BMS.tsx & outline start

Just starting to roughly work on the "view" part of the file and some rough outline componenets.

for-a-greener-dawn committed:
additional outlining

Styling for the feature is further roughed out with most of the variables being declared in the IState interface. The styling is definately bloated, but I'd like advising on where to go from here, as there are properties of the C# software that I'm unsure of in replicating the layout of the feature.

for-a-greener-dawn committed:
Update Power&BMS.tsx

for-a-greener-dawn committed:
Begin Render Setup

Mostly because of an inability to find useful resources(at the moment), for now I've resorted to brute-forcing the render.

for-a-greener-dawn committed:
Is now visually testable

after a bunch of document diving and looking through yarn's website, I figured out how to get Power on the application. Will likely need help on setting up instance-able functions for rendering all the buttons.

for-a-greener-dawn committed:
Added Togglable btns for First Array

Added a second array to handle the instancing of each button's individual functionality. Also mostly finished the amperage readouts. Next steps will be the button array and figuring out how to export/change new variable states with each button click.

for-a-greener-dawn committed:
Finished Mostly Polished Outline

The remaining work on the feature can be almost exclusively backend work now. I will require help incorporating rovecomm functionality

for-a-greener-dawn committed:
Merge branch 'dev' into feature/Power&BMS

for-a-greener-dawn committed:
Began total rework (won't take long)

In the midst of learning how to integrate rovecomm data, it was determined 500 lines of code was too long for this feature. This commit contains the beginning of an array based data structure that should correctly integrate rovecomm data via event listeners. Current direction is pretty solid and should be fairly quick to implement.

for-a-greener-dawn committed:
Update Power&BMS.tsx

for-a-greener-dawn committed:
Bulk Setup 90% Done

The majority of the typing should be done, and I put in placeholders for things I either need help with or am saving for another day.

for-a-greener-dawn committed:
Fixed indenting and const declares

the auto code optimizer suggested using const instead of let for the various "this.state"'s and I fixed some ugly indenting done by the style optimizer

for-a-greener-dawn committed:
Fixed Template Literal Placeholder

Fixed an issue where I tried to insert a variable into a string and needed to use back-tick quotes instead of double quotes. Also removed the "START LOG" placeholder button.

for-a-greener-dawn committed:
Added Some Descriptive Documentation

Just added a few comments describing potentially confusing blocks of code, such as a parent and child div both having the same CSS style

for-a-greener-dawn committed:
Polished Functions and Started Rework on CSS

for-a-greener-dawn committed:
Close to Final Functionality

All that is left is incorporating color background and cleaning up code for readability

for-a-greener-dawn committed:
Merge branch 'dev' into feature/Power&BMS

for-a-greener-dawn committed:
Revert "Merge branch 'dev' into feature/Power&BMS"

This reverts commit 1fd26c8bafe9ab3460e35d9469c7ae78a30fb281, reversing
changes made to 54cc9f182f45c0f7ec60261c48fd0de088cc5679.

for-a-greener-dawn committed:
General Optimizing

for-a-greener-dawn committed:
Revert "Revert "Merge branch 'dev' into feature/Power&BMS""

This reverts commit 98a1c63b48b915798e553c4a11f21530a88aca74.

for-a-greener-dawn committed:
Finalizing Initial Merge of dev into Power

for-a-greener-dawn committed:
Power and BMS Ready for Pull Request

Still need to incorporate color change for readout displays but as it is not crucial for the feature's function, it will be worked on later

thomas-gleiforst committed:
Merge dev into Power&BMS

for-a-greener-dawn committed:
Fixed Errors that Caused Crashes

changed vacuum call and turnOffReboot function so that they don't cause crashes

thomas-gleiforst commented on Electron/src/RED/components/Power&BMS.tsx:
```diff
@@ -0,0 +1,472 @@
+    rovecomm.on("MotorBusCurrent", (data: number) => this.motorBusCurrents(data))
+    rovecomm.on("MotorBusEnabled", (data: number) => this.motorBusEnabled(data))
+    rovecomm.on("SteeringMotorCurrents", (data: number) => this.steeringMotorCurrents(data)) // potential removal
+    rovecomm.on("12VActBusEnable", (data: number) => this.twelveVActBusEnable(data))
+    rovecomm.on("12VLogicBusEnable", (data: number) => this.twelveVLogicBusEnable(data))
```
Update these to "Enabled"

for-a-greener-dawn committed:
Enacted Requested Change

Changed 12VActBusEnable to 12VActBusEnabled and the other one. For continuity, I also changed the function names to reflect the rovecomm listeners

for-a-greener-dawn committed:
Encorporated Changing Background Colors for Readouts

for-a-greener-dawn committed:
Merge dev into Power

for-a-greener-dawn committed:
Added bitmaskUnpack Function

designed to add leading 0's to the bitmask after leading zeros get destoryed by number conversion

for-a-greener-dawn committed:
Finalized Code Ready for Review

Planning to separate the bitmask unpacker to an independent file

for-a-greener-dawn committed:
Fixed small syntax issue

for-a-greener-dawn committed:
Added Independent File for BitmaskUnpack

for-a-greener-dawn committed:
Cleaned Up Visuals and Completed Thorough Bug Test

thomas-gleiforst commented on .vscode/launch.json:
```diff
@@ -0,0 +1,27 @@
+{
```
All these .vscode files should probably be gitignored

thomas-gleiforst commented on Electron/RED/Routes.tsx:
```diff
@@ -0,0 +1,15 @@
+      <Route path={routes.HOME} component={Power} />
+    </Switch>
+  )
+}
```
This file can be deleted

thomas-gleiforst commented on Electron/src/RED/components/Power&BMS.tsx:
```diff
@@ -0,0 +1,548 @@
+}
+
+function turnOffReboot(time: number): void {
+  rovecomm.sendCommand("BMSStop", [time])
```
Probably leave a comment explaining how this functions, as in how time = 5 reboots in five seconds and time = 0 never reboots

thomas-gleiforst commented on Electron/src/RED/components/Power&BMS.tsx:
```diff
@@ -0,0 +1,548 @@
+  constructor(props: IProps) {
+    super(props)
+    this.state = {
+      boardTelemetry: {
+        "Drive LF": { enabled: false, value: 0 },
+        "Drive LR": { enabled: false, value: 0 },
+        "Drive RF": { enabled: false, value: 0 },
+        "Drive RR": { enabled: false, value: 0 },
+        "Steering LF": { value: 0 },
+        "Steering LR": { value: 0 },
+        "Steering RF": { value: 0 },
+        "Steering RR": { value: 0 },
+        "Spare Motor": { enabled: false, value: 0 },
+        Gimbal: { enabled: false, value: 0 },
+        Multimedia: { enabled: false, value: 0 },
+        Autonomy: { enabled: false, value: 0 },
+        Logic: { value: 0 },
+        "12V": { enabled: false, value: 0 },
+        Comms: { enabled: false, value: 0 },
+        Auxiliary: { enabled: false, value: 0 },
+        Drive: { enabled: false, value: 0 },
+        Vacuum: { enabled: false, value: 0 },
+        Nav: { enabled: false },
+        Cameras: { enabled: false },
+        Extra: { enabled: false },
+      },
+      batteryTelemetry: {
+        Temp: { value: 0 },
+        TotalPackCurrent: { value: 0 },
+        TotalPackVoltage: { value: 0 },
+        "Cell 1": { value: 0 },
+        "Cell 2": { value: 0 },
+        "Cell 3": { value: 0 },
+        "Cell 4": { value: 0 },
+        "Cell 5": { value: 0 },
+        "Cell 6": { value: 0 },
+        "Cell 7": { value: 0 },
+        "Cell 8": { value: 0 },
+      },
+    }
```
Consider doing this programmatically like in RON. As in you have several lists that would all default to value = 0, and some value=0 enabled=false, and you could add to the object using those lists instead of having this massive default object defined manually. I don't know if that is worth it or cleaner or not, but

thomas-gleiforst commented on Electron/src/RED/components/Power&BMS.tsx:
```diff
@@ -0,0 +1,548 @@
+      boardTelemetry[motors[i]].value = data[i]
+    }
+    this.setState({ boardTelemetry })
+  }
```
These two functions are essentially repeated over and over with different const lists or with different objects (boardTelemetry or batteryTelemetry) --> you could probably make this one function that takes those in as parameters.

thomas-gleiforst replied:
Brining this down to just 2 functions would make thorough comments much more feasible.

thomas-gleiforst commented on Electron/src/RED/components/Power&BMS.tsx:
```diff
@@ -0,0 +1,548 @@
+
+  motorBusEnabled(data: number[]): void {
+    const motors = ["Drive LF", "Drive LR", "Drive RF", "Drive RR", "Spare Motor"]
+    const bitmask = BitmaskUnpack(data[0], motors)
```
You could just pass in the length of motors instead of the motors list itself

thomas-gleiforst commented on Electron/src/RED/components/Power&BMS.tsx:
```diff
@@ -0,0 +1,548 @@
+    this.setState({ batteryTelemetry })
+  }
+
+  buttonToggle(bus: string): void {
```
Again, add some more comments to when this is being called and what it does. Mentioning that this is for enable buttons will make it more clear if someone attempts to reuse it.

thomas-gleiforst commented on Electron/src/RED/components/Power&BMS.tsx:
```diff
@@ -0,0 +1,548 @@
+          },
+        },
+      },
+      () => this.packCommand(bus)
```
While I agree you need to use the setState function callback here, its always a weird thing to do and you should add a comment as to why

thomas-gleiforst commented on Electron/src/RED/components/Power&BMS.tsx:
```diff
@@ -0,0 +1,548 @@
+  }
+
+  allMotorToggle(button: boolean): void {
+    const motors = ["Drive LF", "Drive LR", "Drive RF", "Drive RR", "Spare Motor"]
```
If you are going to constrict the 12 functions for each enable and current read in down to just 2 functions with these lists passed in, then these lists will be global constants, and you should use that instead of redeclaring all the time.

thomas-gleiforst commented on Electron/src/RED/components/Power&BMS.tsx:
```diff
@@ -0,0 +1,548 @@
+        }
+      }
+    }
+    this.setState({ boardTelemetry }, () => this.packCommand(motors[0]))
```
Again, using the callback is weird, leave a comment, and explain why you pass in motors[0]

thomas-gleiforst commented on Electron/src/RED/components/Power&BMS.tsx:
```diff
@@ -0,0 +1,548 @@
+    const motors = ["Drive LF", "Drive LR", "Drive RF", "Drive RR", "Spare Motor"]
+    const actBus = ["Gimbal", "Multimedia", "Auxiliary"]
+    const logicBus = ["Gimbal", "Multimedia", "Autonomy", "Drive", "Nav", "Cameras", "Extra"]
+    const thirtyVBus = ["12V", "Comms", "Auxiliary", "Drive"]
```
Mentioned previously, but all these lists are used many times and should be a global constant

thomas-gleiforst commented on Electron/src/RED/components/Power&BMS.tsx:
```diff
@@ -0,0 +1,548 @@
+      motors.forEach(motor => {
+        newBitMask += this.state.boardTelemetry[motor].enabled ? "1" : "0"
+      })
+      rovecomm.sendCommand("MotorBusEnable", [parseInt(newBitMask, 2)])
```
In at least this first case you should comment what is happening here as specifically as possible.

thomas-gleiforst commented on Electron/src/RED/components/Power&BMS.tsx:
```diff
@@ -0,0 +1,548 @@
+    }
+    if (actBus.includes(bus)) {
+      let newBitMask = ""
+      actBus.forEach(motor => {
```
don't call all of these motor

thomas-gleiforst commented on Electron/src/RED/components/Power&BMS.tsx:
```diff
@@ -0,0 +1,548 @@
+    const actBus = ["Gimbal", "Multimedia", "Auxiliary"]
+    const logicBus = ["Gimbal", "Multimedia", "Autonomy", "Drive", "Nav", "Cameras", "Extra"]
+    const thirtyVBus = ["12V", "Comms", "Auxiliary", "Drive"]
+    if (motors.includes(bus)) {
```
These are a bunch of different ifs so that something can appear in multiple lists and be toggled together. But we may want to be toggled independently, which would be a good future feature - the UI would be probably the hardest part of this consideration.

thomas-gleiforst commented on Electron/src/RED/components/Power&BMS.tsx:
```diff
@@ -0,0 +1,548 @@
+            </div>
+            <div
+              style={{ ...column, flexGrow: 1 }}
+              /* second column of buttons and readouts */
+            >
+              {[
+                "Gimbal",
+                "Multimedia",
+                "Autonomy",
+                "Logic", // doesn't actually have enable
+                "12V",
+                "Comms",
+                "Auxiliary",
+                "Drive",
+                "Vacuum",
+              ].map(part => {
+                return (
+                  <div key={part} style={row}>
+                    {"enabled" in this.state.boardTelemetry[part] ? (
+                      <button type="button" onClick={() => this.buttonToggle(part)} style={btnStyle}>
+                        {this.state.boardTelemetry[part].enabled ? "Enabled" : "Disabled"}
+                      </button>
+                    ) : (
+                      <div style={{ width: "70px", backgroundColor: "hsl(0, 0%, 90%)" }} />
+                    )}
+                    <div style={ColorStyleConverter(this.state.boardTelemetry[part].value, 0, 7, 15, 120, 0, readout)}>
+                      <h3 style={textPad}>{part}</h3>
+                      <h3 style={textPad}>
+                        {this.state.boardTelemetry[part].value.toLocaleString(undefined, {
+                          minimumFractionDigits: 1,
+                          maximumFractionDigits: 1,
+                          minimumIntegerDigits: 2,
+                        })}
+                        &nbsp;A
+                      </h3>
+                    </div>
+                  </div>
+                )
+              })}
+            </div>
```
you could potentially make a map function to map over these two map functions....

thomas-gleiforst replied:
You can look at pingtool lines 196-241 to see what I mean where you pass in the keys for the map from another map

thomas-gleiforst commented on Electron/src/RED/components/Power&BMS.tsx:
```diff
@@ -0,0 +1,548 @@
+                    ) : (
+                      <div style={{ width: "70px", backgroundColor: "hsl(0, 0%, 90%)" }} />
+                    )}
+                    <div style={ColorStyleConverter(this.state.boardTelemetry[motor].value, 0, 7, 15, 120, 0, readout)}>
```
Consider making global constants for the cutoff, max, min, lowHues, and HighHue

thomas-gleiforst commented on Electron/src/RED/components/Power&BMS.tsx:
```diff
@@ -0,0 +1,548 @@
+                          maximumFractionDigits: 1,
+                          minimumIntegerDigits: 2,
+                        })}
+                        &nbsp;A
```
Use a template string instead \`${this.state.boardTelemetry[motor].value.toLocaleString(undefined, {
                          minimumFractionDigits: 1,
                          maximumFractionDigits: 1,
                          minimumIntegerDigits: 2,
                        })} A\`

Commit:
Merge branch 'dev' into feature/Power&BMS

for-a-greener-dawn committed:
Function Conglomeration and butmaskunpack de-specification

for-a-greener-dawn committed:
Merge branch 'feature/Power&BMS' of https://github.com/MissouriMRDT/BaseStation_Software into feature/Power&BMS

for-a-greener-dawn committed:
Requested Changes Completed

The last thing to do is just some short documentation I'm too tired to type out at the moment

for-a-greener-dawn committed:
Added more requested documentation

thomas-gleiforst committed:
Gitignore and remove a few more files
